[Flaky Test] [sig-api-machinery] TestStreamTranslator_WebSocketServerErrors raises error in unrelated PR

### Which jobs are flaking?

pull-kubernetes-unit

### Which tests are flaking?

TestStreamTranslator_WebSocketServerErrors

### Since when has it been flaking?

2025-06-24 15:33:39 +0000 UTC

### Testgrid link

_No response_

### Reason for failure (if possible)

```log
=== RUN   TestStreamTranslator_WebSocketServerErrors
    streamtranslator_test.go:766: +# HELP apiserver_stream_translator_requests_total [ALPHA] Total number of requests that were handled by the StreamTranslatorProxy, which processes streaming RemoteCommand/V5
        +# TYPE apiserver_stream_translator_requests_total counter
        +apiserver_stream_translator_requests_total{code="400"} 1
         
--- FAIL: TestStreamTranslator_WebSocketServerErrors (0.01s)
✖  k8s.io/apiserver/pkg/util/proxy (2.744s)
✓  k8s.io/apiserver/pkg/util/webhook (2.891s)
✓  k8s.io/apiserver/pkg/util/shufflesharding (8.339s)
✓  k8s.io/apiserver/pkg/util/x509metrics (1.054s)
✓  k8s.io/apiserver/pkg/validation (1.032s)
```

```log
=== SKIP: pkg/volume TestSetVolumeOwnershipOwner (0.00s)
    volume_linux_test.go:501: running as non-root
 === Failed
=== FAIL: k8s.io/apiserver/pkg/util/proxy TestStreamTranslator_WebSocketServerErrors (0.01s)
    streamtranslator_test.go:766: +# HELP apiserver_stream_translator_requests_total [ALPHA] Total number of requests that were handled by the StreamTranslatorProxy, which processes streaming RemoteCommand/V5
        +# TYPE apiserver_stream_translator_requests_total counter
        +apiserver_stream_translator_requests_total{code="400"} 1
         
 DONE 72021 tests, 204 skipped, 1 failure in 987.573s
processing junit xml file : /logs/artifacts/junit_20250624-153346.xml
done.
make: *** [Makefile:192: test] Error 1
```

### Anything else we need to know?

PR link: https://github.com/kubernetes/kubernetes/pull/132484
Job ran flaky test: https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/132484/pull-kubernetes-unit/1937534214664622080

### Relevant SIG(s)

/sig api-machinery

> UT flake in the same file:
> 
> * [k8s.io/apiserver/pkg/util: proxy TestStreamTranslator_TTYResizeChannel failing in `pull-kubernetes-unit` #130018](https://github.com/kubernetes/kubernetes/issues/130018)
> * [Failure cluster [6bb013b4...] #131289](https://github.com/kubernetes/kubernetes/issues/131289)
> * [Failure cluster [07f9cb53...]: TestStreamTranslator_MultipleReadChannels #131591](https://github.com/kubernetes/kubernetes/issues/131591)
> * [[Flaking Test] k8s.io/apiserver/pkg/util/proxy.TestStreamTranslator #132279](https://github.com/kubernetes/kubernetes/issues/132279)
> 
> The flake reason seems to be different, but all in https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/util/proxy/streamtranslator_test.go

looking into it, indeed it seems something that impacts all tests

>     streamtranslator_test.go:170: error on createHTTPStreams: unable to upgrade: X-Stream-Protocol-Version is required

>   streamtranslator_test.go:702:  # HELP apiserver_stream_translator_requests_total [ALPHA] Total number of requests that were handled by the StreamTranslatorProxy, which processes streaming RemoteCommand/V5
         # TYPE apiserver_stream_translator_requests_total counter
         apiserver_stream_translator_requests_total{code="200"} 1
        -apiserver_stream_translator_requests_total{code="400"} 1

> === RUN   TestStreamTranslator_WebSocketServerErrors
    streamtranslator_test.go:766: +# HELP apiserver_stream_translator_requests_total [ALPHA] Total number of requests that were handled by the StreamTranslatorProxy, which processes streaming RemoteCommand/V5
        +# TYPE apiserver_stream_translator_requests_total counter
        +apiserver_stream_translator_requests_total{code="400"} 1

repro with the usual steps 
```
/tmp/go-stress-20250701T150448-3568964630
E0701 15:05:05.175438 2646701 conn.go:339] Error on socket receive: read tcp 127.0.0.1:35483->127.0.0.1:36840: use of closed network connection
E0701 15:05:05.315904 2646701 conn.go:339] Error on socket receive: read tcp 127.0.0.1:44005->127.0.0.1:33950: use of closed network connection
E0701 15:05:05.462895 2646701 conn.go:339] Error on socket receive: read tcp 127.0.0.1:41135->127.0.0.1:36816: use of closed network connection
E0701 15:05:06.967677 2646701 conn.go:339] Error on socket receive: read tcp 127.0.0.1:33633->127.0.0.1:40844: use of closed network connection
E0701 15:05:07.877294 2646701 conn.go:339] Error on socket receive: read tcp 127.0.0.1:39161->127.0.0.1:59226: use of closed network connection
E0701 15:05:07.936521 2646701 websocket.go:297] Unknown stream id 1, discarding message
E0701 15:05:07.955504 2646701 conn.go:339] Error on socket receive: read tcp 127.0.0.1:35727->127.0.0.1:44980: use of closed network connection
E0701 15:05:07.955983 2646701 websocket.go:297] Unknown stream id 1, discarding message
E0701 15:05:07.956029 2646701 websocket.go:297] Unknown stream id 3, discarding message
E0701 15:05:07.980157 2646701 websocket.go:297] Unknown stream id 1, discarding message
E0701 15:05:08.004096 2646701 websocket.go:297] Unknown stream id 1, discarding message
E0701 15:05:08.005565 2646701 conn.go:339] Error on socket receive: read tcp 127.0.0.1:42065->127.0.0.1:40834: use of closed network connection
E0701 15:05:08.023715 2646701 websocket.go:297] Unknown stream id 1, discarding message
--- FAIL: TestStreamTranslator_BlockRedirects (0.12s)
    streamtranslator_test.go:828: expected redirect not allowed error, got (closed all streams)
E0701 15:05:08.208695 2646701 streamtunnel.go:367] SPDY upgrade failed: 500 Internal Server Error
E0701 15:05:08.240833 2646701 streamtunnel.go:367] SPDY upgrade failed: 502 Bad Gateway
E0701 15:05:08.264321 2646701 streamtunnel.go:367] SPDY upgrade failed: 503 Service Unavailable
E0701 15:05:08.283448 2646701 streamtunnel.go:153] Hijack called after write
E07
…
20s: 167 runs so far, 1 failures (0.60%
```

--- a/staging/src/k8s.io/apiserver/pkg/util/proxy/streamtranslator.go
+++ b/staging/src/k8s.io/apiserver/pkg/util/proxy/streamtranslator.go
@@ -73,14 +73,14 @@ func (h *StreamTranslatorHandler) ServeHTTP(w http.ResponseWriter, req *http.Req
 	// Creating SPDY executor, ensuring redirects are not followed.
 	spdyRoundTripper, err := spdy.NewRoundTripperWithConfig(spdy.RoundTripperConfig{UpgradeTransport: h.Transport, PingPeriod: 5 * time.Second})
 	if err != nil {
-		websocketStreams.writeStatus(apierrors.NewInternalError(err)) //nolint:errcheck
 		metrics.IncStreamTranslatorRequest(req.Context(), strconv.Itoa(http.StatusInternalServerError))
+		websocketStreams.writeStatus(apierrors.NewInternalError(err)) //nolint:errcheck
 		return
 	}
 	spdyExecutor, err := remotecommand.NewSPDYExecutorRejectRedirects(spdyRoundTripper, spdyRoundTripper, "POST", h.Location)
 	if err != nil {
-		websocketStreams.writeStatus(apierrors.NewInternalError(err)) //nolint:errcheck
 		metrics.IncStreamTranslatorRequest(req.Context(), strconv.Itoa(http.StatusInternalServerError))
+		websocketStreams.writeStatus(apierrors.NewInternalError(err)) //nolint:errcheck
 		return
 	}
 
@@ -121,27 +121,27 @@ func (h *StreamTranslatorHandler) ServeHTTP(w http.ResponseWriter, req *http.Req
 	if err != nil {
 		//nolint:errcheck   // Ignore writeStatus returned error
 		if statusErr, ok := err.(*apierrors.StatusError); ok {
-			websocketStreams.writeStatus(statusErr)
 			// Increment status code returned within status error.
 			metrics.IncStreamTranslatorRequest(req.Context(), strconv.Itoa(int(statusErr.Status().Code)))
+			websocketStreams.writeStatus(statusErr)
 		} else if exitErr, ok := err.(exec.CodeExitError); ok && exitErr.Exited() {
-			websocketStreams.writeStatus(codeExitToStatusError(exitErr))
 			// Returned an exit code from the container, so not an error in
 			// stream translator--add StatusOK to metrics.
 			metrics.IncStreamTranslatorRequest(req.Context(), strconv.Itoa(http.StatusOK))
+			websocketStreams.writeStatus(codeExitToStatusError(exitErr))
 		} else {
-			websocketStreams.writeStatus(apierrors.NewInternalError(err))
 			metrics.IncStreamTranslatorRequest(req.Context(), strconv.Itoa(http.StatusInternalServerError))
+			websocketStreams.writeStatus(apierrors.NewInternalError(err))
 		}
 		return
 	}
 
+	metrics.IncStreamTranslatorRequest(req.Context(), strconv.Itoa(http.StatusOK))
 	// Write the success status back to the WebSocket client.
 	//nolint:errcheck
 	websocketStreams.writeStatus(&apierrors.StatusError{ErrStatus: metav1.Status{
 		Status: metav1.StatusSuccess,
 	}})
-	metrics.IncStreamTranslatorRequest(req.Context(), strconv.Itoa(http.StatusOK))
 }
 
 // translatorSizeQueue feeds the size events from the WebSocket

--- a/staging/src/k8s.io/apiserver/pkg/util/proxy/streamtranslator_test.go
+++ b/staging/src/k8s.io/apiserver/pkg/util/proxy/streamtranslator_test.go
@@ -755,15 +755,21 @@ func TestStreamTranslator_WebSocketServerErrors(t *testing.T) {
 			t.Errorf("expected websocket bad handshake error, got (%s)", err)
 		}
 	}
-	// Validate the streamtranslator metrics; should have one 500 failure.
+	// Validate the streamtranslator metrics; should have one 400 failure.
+	// Use polling to wait for the metric to be updated asynchronously.
 	metricNames := []string{"apiserver_stream_translator_requests_total"}
 	expected := `
 # HELP apiserver_stream_translator_requests_total [ALPHA] Total number of requests that were handled by the StreamTranslatorProxy, which processes streaming RemoteCommand/V5
 # TYPE apiserver_stream_translator_requests_total counter
 apiserver_stream_translator_requests_total{code="400"} 1
 `
-	if err := testutil.GatherAndCompare(legacyregistry.DefaultGatherer, strings.NewReader(expected), metricNames...); err != nil {
-		t.Fatal(err)
+	if err := wait.PollUntilContextTimeout(context.Background(), 100*time.Millisecond, 2*time.Second, true, func(ctx context.Context) (bool, error) {
+		if testutil.GatherAndCompare(legacyregistry.DefaultGatherer, strings.NewReader(expected), metricNames...) == nil {
+			return true, nil
+		}
+		return false, nil
+	}); err != nil {
+		t.Fatalf("Failed to observe metric after waiting 2 seconds: %v", err)
 	}
 }
 

--- a/staging/src/k8s.io/apiserver/pkg/util/proxy/streamtranslator_test.go
+++ b/staging/src/k8s.io/apiserver/pkg/util/proxy/streamtranslator_test.go
@@ -771,8 +771,8 @@ apiserver_stream_translator_requests_total{code="400"} 1
 // redirects; it will thrown an error instead.
 func TestStreamTranslator_BlockRedirects(t *testing.T) {
 	metrics.Register()
-	metrics.ResetForTest()
 	t.Cleanup(metrics.ResetForTest)
+
 	for _, statusCode := range []int{
 		http.StatusMovedPermanently,  // 301
 		http.StatusFound,             // 302
@@ -780,65 +780,67 @@ func TestStreamTranslator_BlockRedirects(t *testing.T) {
 		http.StatusTemporaryRedirect, // 307
 		http.StatusPermanentRedirect, // 308
 	} {
-		// Create upstream fake SPDY server which returns a redirect.
-		spdyServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
-			w.Header().Set("Location", "/")
-			w.WriteHeader(statusCode)
-		}))
-		defer spdyServer.Close()
-		spdyLocation, err := url.Parse(spdyServer.URL)
-		if err != nil {
-			t.Fatalf("Unable to parse spdy server URL: %s", spdyServer.URL)
-		}
-		spdyTransport, err := fakeTransport()
-		if err != nil {
-			t.Fatalf("Unexpected error creating transport: %v", err)
-		}
-		streams := Options{Stdout: true}
-		streamTranslator := NewStreamTranslatorHandler(spdyLocation, spdyTransport, 0, streams)
-		streamTranslatorServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
-			streamTranslator.ServeHTTP(w, req)
-		}))
-		defer streamTranslatorServer.Close()
-		// Now create the websocket client (executor), and point it to the "streamTranslatorServer".
-		streamTranslatorLocation, err := url.Parse(streamTranslatorServer.URL)
-		if err != nil {
-			t.Fatalf("Unable to parse StreamTranslator server URL: %s", streamTranslatorServer.URL)
-		}
-		exec, err := remotecommand.NewWebSocketExecutor(&rest.Config{Host: streamTranslatorLocation.Host}, "GET", streamTranslatorServer.URL)
-		if err != nil {
-			t.Errorf("unexpected error creating websocket executor: %v", err)
-		}
-		errorChan := make(chan error)
-		go func() {
-			// Start the streaming on the WebSocket "exec" client.
-			// Should return "redirect not allowed" error.
-			errorChan <- exec.StreamWithContext(context.Background(), remotecommand.StreamOptions{})
-		}()
+		t.Run(fmt.Sprintf("statusCode=%d", statusCode), func(t *testing.T) {
+			metrics.ResetForTest()
+			// Create upstream fake SPDY server which returns a redirect.
+			spdyServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+				w.Header().Set("Location", "/")
+				w.WriteHeader(statusCode)
+			}))
+			defer spdyServer.Close()
+			spdyLocation, err := url.Parse(spdyServer.URL)
+			if err != nil {
+				t.Fatalf("Unable to parse spdy server URL: %s", spdyServer.URL)
+			}
+			spdyTransport, err := fakeTransport()
+			if err != nil {
+				t.Fatalf("Unexpected error creating transport: %v", err)
+			}
+			streams := Options{Stdout: true}
+			streamTranslator := NewStreamTranslatorHandler(spdyLocation, spdyTransport, 0, streams)
+			streamTranslatorServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+				streamTranslator.ServeHTTP(w, req)
+			}))
+			defer streamTranslatorServer.Close()
+			// Now create the websocket client (executor), and point it to the "streamTranslatorServer".
+			streamTranslatorLocation, err := url.Parse(streamTranslatorServer.URL)
+			if err != nil {
+				t.Fatalf("Unable to parse StreamTranslator server URL: %s", streamTranslatorServer.URL)
+			}
+			exec, err := remotecommand.NewWebSocketExecutor(&rest.Config{Host: streamTranslatorLocation.Host}, "GET", streamTranslatorServer.URL)
+			if err != nil {
+				t.Errorf("unexpected error creating websocket executor: %v", err)
+			}
+			errorChan := make(chan error)
+			go func() {
+				// Start the streaming on the WebSocket "exec" client.
+				// Should return "redirect not allowed" error.
+				errorChan <- exec.StreamWithContext(context.Background(), remotecommand.StreamOptions{})
+			}()
 
-		select {
-		case <-time.After(wait.ForeverTestTimeout):
-			t.Fatalf("expect stream to be closed after connection is closed.")
-		case err := <-errorChan:
-			// Must return "redirect now allowed" error.
-			if err == nil {
-				t.Fatalf("expected error, but received none")
+			select {
+			case <-time.After(wait.ForeverTestTimeout):
+				t.Fatalf("expect stream to be closed after connection is closed.")
+			case err := <-errorChan:
+				// Must return "redirect now allowed" error.
+				if err == nil {
+					t.Fatalf("expected error, but received none")
+				}
+				if !strings.Contains(err.Error(), "redirect not allowed") {
+					t.Errorf("expected redirect not allowed error, got (%s)", err)
+				}
 			}
-			if !strings.Contains(err.Error(), "redirect not allowed") {
-				t.Errorf("expected redirect not allowed error, got (%s)", err)
-			}
-		}
-		// Validate the streamtranslator metrics; should have one 500 failure each loop.
-		metricNames := []string{"apiserver_stream_translator_requests_total"}
-		expected := `
+			// Validate the streamtranslator metrics; should have one 500 failure each loop.
+			metricNames := []string{"apiserver_stream_translator_requests_total"}
+			expected := `
 # HELP apiserver_stream_translator_requests_total [ALPHA] Total number of requests that were handled by the StreamTranslatorProxy, which processes streaming RemoteCommand/V5
 # TYPE apiserver_stream_translator_requests_total counter
 apiserver_stream_translator_requests_total{code="500"} 1
 `
-		if err := testutil.GatherAndCompare(legacyregistry.DefaultGatherer, strings.NewReader(expected), metricNames...); err != nil {
-			t.Fatal(err)
-		}
-		metrics.ResetForTest() // Clear metrics each loop
+			if err := testutil.GatherAndCompare(legacyregistry.DefaultGatherer, strings.NewReader(expected), metricNames...); err != nil {
+				t.Fatal(err)
+			}
+		})
 	}
 }
 

